src/tz_macos.rs

```diff
@@ -1,22 +1,50 @@
-use core_foundation_sys::base::{CFRelease, CFTypeRef};
-use core_foundation_sys::string::{kCFStringEncodingUTF8, CFStringGetCStringPtr};
+use core_foundation_sys::base::{Boolean, CFRange, CFRelease, CFTypeRef};
+use core_foundation_sys::string::{kCFStringEncodingUTF8, CFStringGetBytes, CFStringGetLength};
 use core_foundation_sys::timezone::{CFTimeZoneCopySystem, CFTimeZoneGetName};
 
 pub(crate) fn get_timezone_inner() -> Result<String, crate::GetTimezoneError> {
-    unsafe {
-        Dropping::new(CFTimeZoneCopySystem())
-            .and_then(|tz| Dropping::new(CFTimeZoneGetName(tz.0)))
-            .and_then(|name| {
-                let name = CFStringGetCStringPtr(name.0, kCFStringEncodingUTF8);
-                if name.is_null() {
-                    None
-                } else {
-                    Some(name)
-                }
-            })
-            .and_then(|name| std::ffi::CStr::from_ptr(name).to_str().ok())
-            .map(|name| name.to_owned())
-            .ok_or(crate::GetTimezoneError::OsError)
+    unsafe { get_timezone().ok_or(crate::GetTimezoneError::OsError) }
+}
+
+#[inline]
+unsafe fn get_timezone() -> Option<String> {
+    // The longest name in the IANA time zone database is 25 ASCII characters long.
+    const MAX_LEN: usize = 32;
+
+    // Get system time zone, and borrow its name.
+    let tz = Dropping::new(CFTimeZoneCopySystem())?;
+    let name = match CFTimeZoneGetName(tz.0) {
+        name if !name.is_null() => name,
+        _ => return None,
+    };
+
+    // Copy the name into the buffer.
+    let mut buf = [0; MAX_LEN];
+    let mut buf_bytes = 0;
+    let range = CFRange {
+        location: 0,
+        length: CFStringGetLength(name),
+    };
+    if CFStringGetBytes(
+        name,
+        range,
+        kCFStringEncodingUTF8,
+        b'\0',
+        false as Boolean,
+        buf.as_mut_ptr(),
+        buf.len() as isize,
+        &mut buf_bytes,
+    ) != range.length
```

Comment:
This return value is the number of encoding-aware characters, not the length of the byte string.

the length stored in the out parameter `buf_bytes` is the number of bytes copied.

https://developer.apple.com/documentation/corefoundation/1543006-cfstringgetbytes

Reply:
Yes, I check that all characters in the original encoding were consumed, not more or less. But I use `buf_bytes` to build the UTF-8 string.

core-foundation even asserts that the length matches: <https://docs.rs/core-foundation/0.9.3/src/core_foundation/string.rs.html#82>

Reply:
Ahh I see.

https://developer.apple.com/documentation/corefoundation/1543006-cfstringgetbytes

> ## Return Value
>
> The number of characters converted.